Debugged indexing errors

AlexanderEvans opened the pull request:
Manged to debug it and retain null termination.  Also made the API a bit safer to use(though less shorthands are now available).  If this looks good please merge it down.  :)  I'll also go ahead and change the repo config quick to require code review from at least one other member.

Commit:
Debugged indexing errors

Manged to debug it and retain null termination.  Also made the api a bit safer to use(though less shorthands are now available).

Commit:
Debugged replacement

Replacement was able to recurse infinitely and failed to shrink it the replacement string was shorter than the old substring.  These issues were fixed.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -25,6 +27,28 @@ public NativeString Concatenate(NativeString A, NativeString B)
+        {
+            return this + B;
+        }
+
```
I personally would argue that having the operator overloads call Append (Append is concatenate in C# speak) instead of the other way around.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -33,8 +57,8 @@ public NativeString Concatenate(NativeString A, NativeString B)
-            EnforceNullTermination(ref B);
+            A.RemoveNullTermination();
+            B.EnforceNullTermination();
 
```
Get rid of null terminators. They waste CPU cycles and occur an unnecessary cache miss for every method call.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -47,6 +71,50 @@ public NativeString Concatenate(NativeString A, NativeString B)
+        }
+
+        public NativeString Add(char B)
+        {
```
Probably should rename everything named Add to Append to match StringBuilder API.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -79,6 +147,16 @@ public override bool Equals(object obj)
+        {
+            return nativeString;
+        }
+
```
We don't need static method copies for everything. At this point let's just get all the core API as member methods. And then we can write extensions for useful features. But right now this right here makes no sense.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -97,17 +175,52 @@ public static bool Equals(NativeString A, NativeString B)
+
+            return A;
+        }
 
```
ResizeUninitialized does not work here if the new string is smaller. Use clear and then append.

AlexanderEvans replied:
Why doesn't this work?  It does shrink the length, just not the capacity.  Isn't that what we want?  Then we jut loop through and set everything?

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -97,17 +175,52 @@ public static bool Equals(NativeString A, NativeString B)
         }
 
         public static bool Replace(ref NativeString A, NativeString oldString, NativeString replacement)
         {
```
You don't need ref for NativeContainers. They aren't RAII like STL. They are basically like pointers, which is why they are structs. (Actually the struct literally contains a pointer to the actual data.)

AlexanderEvans replied:
Ah, ok!  This explains some... well, what I perceived as, very weird behavior as i was writing this script.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -247,10 +437,14 @@ public static void OverWrite(ref NativeString A, int position, NativeString stam
             }
         }
 
```
I think on Friday I will write some wrappers so that it is trivial to write replace and insert algorithms without having to allocate new NativeStrings. In the meantime I suggest you focus on getting Append() methods for the different data types (int, float, double, char, byte, short, string, NativeString, ect. And then make + operator overloads for them.

Dreaming381 commented on NativeCharList/Assets/NativeString.cs:
```diff
@@ -247,10 +437,14 @@ public static void OverWrite(ref NativeString A, int position, NativeString stam
+        //{
+        //    return new NativeString(A.Length, Allocator.Persistent);
+        //}
 
```
Remove permanently. Make an overloaded constructor instead.

AlexanderEvans replied:
This is the right solution.  I'll make sure to do that later.

Dreaming381 reviewed (commented):
Lots of comments here. Let's gut all the Inserts and Replace algorithms until I get a chance to build up an internal API that uses UnsafeUtility to make these types of algorithms fast. In the meantime, focus your energy on converting other data types to NativeStrings and appending them (append the characters as you generate them, do not allocate a new NativeString). That is by far the most common use case and once you get that working we can test it out using Jobs and Burst. But remember, no internal allocating!

General Append algorithm:

1. Calculate the number of characters to append.
2. ResizeUnitialized(length + newCharCount)
3. Assign your new characters

If it is not trivial to count the number of characters to append, just append each character and eat the cost of the increment operation for every character added instead of using ResizeUnitialized. Most likely those increment operations will be more efficient than calculating the number of characters to add.

Commit:
optimized Replace by removing intermediate array

Commit:
Fixed bug in replacement

Didn't test it, but there is probably a bug in replacement, and this probably fixes that.

Commit:
removed incorrect license

Dreaming381 commented:
Also wrong branch.
